### PR TITLE
LLVM: Fix allocatable strings

### PR DESCRIPTION
**Pull request by Shaikh-Ubaid**
towards https://github.com/lfortran/lfortran/issues/1953#issuecomment-1694403926

### CODE REVIEW

**Comment by Shaikh-Ubaid** [comment]
This is ready for review.

**Comment by Shaikh-Ubaid** [comment]
We need to update the `fastgpt` fork link/commit-id before merging.

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 1]
```diff
@@ -914,9 +914,6 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
                     llvm::Value* m_len = tmp;
                     std::vector<llvm::Value*> args = {x_arr, m_len};
```
Comment (certik) [inline src/libasr/codegen/asr_to_llvm.cpp:915]: I wonder if you can put `m_len+1` here?

File: src/libasr/runtime/lfortran_intrinsics.c [hunk 2]
```diff
@@ -1409,7 +1409,8 @@ LFORTRAN_API void _lfortran_free(char* ptr) {
 }
 
 LFORTRAN_API void _lfortran_alloc(char** ptr, int32_t size) {
-    *ptr = (char *) malloc(size);
+    *ptr = (char *) malloc(size + 1);
+    (*ptr)[size] = '\0';
 }
```
Comment (certik) [inline src/libasr/runtime/lfortran_intrinsics.c:1414]: I think we need to do exactly this, but only for strings. Not for general allocation.

Comment (Shaikh-Ubaid) [reply src/libasr/runtime/lfortran_intrinsics.c]: Got it.

File: src/libasr/codegen/asr_to_llvm.cpp (hunk 1 above)
Comment (Shaikh-Ubaid) [reply src/libasr/codegen/asr_to_llvm.cpp]: Got it.

**Comment by Shaikh-Ubaid** [comment]
At the moment, this PR contains only two commits related to `fastgpt` version to be used. Since, we use the `size(x)` (instead of `size(x) + 1`) version of `fastgpt` the CI should fail.

**Comment by Shaikh-Ubaid** [comment]
> At the moment, this PR contains only two commits related to `fastgpt` version to be used. Since, we use the `size(x)` (instead of `size(x) + 1`) version of `fastgpt` the CI should fail.

And it seems to fail as expected.

File: .github/workflows/CI.yml [hunk 3]
```diff
@@ -406,6 +406,43 @@ jobs:
...
+              -DCMAKE_BUILD_TYPE=Debug \
+              -DWITH_LLVM=yes \
+              -DLFORTRAN_BUILD_ALL=yes \
+              -DWITH_STACKTRACE=no \
+              -DCMAKE_PREFIX_PATH="$CONDA_PREFIX" \
+              -DCMAKE_INSTALL_PREFIX=`pwd`/inst \
+              -DCMAKE_C_COMPILER_LAUNCHER=sccache \
+              -DCMAKE_CXX_COMPILER_LAUNCHER=sccache
+
+            cmake --build . -j16 --target install
```
Comment (certik) [inline .github/workflows/CI.yml:444]: Can you split this into a separate PR? Given how quickly all these tests run, I was thinking of doing all these 3rd party programs test in one CI test, to save time. Most of the time is spent in Conda and in building LFortran. So after that I would test all our programs. You can rename the test to say 3rd party codes.

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 4]
```diff
@@ -912,11 +912,14 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
...
+                    llvm::Value* alloc_size = builder->CreateAdd(m_len, const_one);
+                    std::vector<llvm::Value*> args = {x_arr, alloc_size};
                     builder->CreateCall(fn, args);
-                    builder->CreateMemSet(LLVM::CreateLoad(*builder, x_arr),
-                        llvm::ConstantInt::get(llvm::Type::getInt8Ty(context), llvm::APInt(8, 0)),
-                        m_len, llvm::MaybeAlign());
+                    std::vector<llvm::Value*> idx_vec = {m_len};
+                    tmp = CreateGEP(CreateLoad(x_arr), idx_vec);
+                    llvm::Value* str_end_null = llvm::ConstantInt::get(llvm::Type::getInt8Ty(context), llvm::APInt(8, 0));
+                    builder->CreateStore(str_end_null, tmp);
```
Comment (certik) [inline src/libasr/codegen/asr_to_llvm.cpp:922]: I think that's fine. Why not leaving the memset there? The reason being that most of the time, the memory will contain 0s (null char) anyway, but sometimes it won't, and we only set the last one here. But that means that `strlen` will return different numbers based on where the null char is.

In general, we should refactor how we handle strings, I think it's a lot of adhoc fixes. But for now, let's just do this "quick fix", but as robustly as we can.

**Comment by certik** [comment]
Great job, I think this approach will work, left some comments.

File: .github/workflows/CI.yml (hunk 3 above)
Comment (Shaikh-Ubaid) [reply .github/workflows/CI.yml:444]: >Can you split this into a separate PR? Given how quickly all these tests run, I was thinking of doing all these 3rd party programs test in one CI test, to save time. Most of the time is spent in Conda and in building LFortran

Sure, I will send it in a separate PR.

>Most of the time is spent in Conda and in building LFortran

I agree. I think one of approaches could be to build LFortran in one of the jobs and use it with other jobs. I think there are github actions that allow us to do this. I think they upload the build file/binary to cloud and then download/use it in other jobs. We might need to investigate about it.

Comment (certik) [reply .github/workflows/CI.yml]: Yes, we should build LFortran once and share. But for now let's just use one CI test for all the 3rd party apps.

File: src/libasr/codegen/asr_to_llvm.cpp (hunk 4 above)
Comment (certik) [reply src/libasr/codegen/asr_to_llvm.cpp:922]: (But if this works, we can change to MemSet in a subsequent PR.)

Comment (Shaikh-Ubaid) [reply src/libasr/codegen/asr_to_llvm.cpp:922]: >In general, we should refactor how we handle strings, I think it's a lot of adhoc fixes. But for now, let's just do this "quick fix", but as robustly as we can.

Is the current change in this PR with respect to setting of null char at the string end fine?

Comment (certik) [reply src/libasr/codegen/asr_to_llvm.cpp:922]: If it passes the CI, then it's fine to merge. I would still feel better with the memset of 0, I think it's more robust, but let's do it after this PR is merged.

**Review by certik — APPROVED** [review]
I think that this is fine, thanks @Shaikh-Ubaid for fixing it!

File: .github/workflows/CI.yml (hunk 3 above)
Comment (Shaikh-Ubaid) [reply .github/workflows/CI.yml]: >I was thinking of doing all these 3rd party programs test in one CI test, to save time

By the way, one of the limitations of testing all 3rd party program in one CI test, I think is that if one the third party program fails at the CI, then the other third party programs that follow it are not tested. Thus we do not get to know which all third party programs fail with a given change (I think this information is usually helpful for debugging as we can look up the 3rd party codes that fail and debug/fix our code/bug).

Comment (certik) [reply .github/workflows/CI.yml]: Sure, although typically if a 3rd party code breaks, and none of our integration_tests breaks, it means we have to isolate the bug and add a dedicated integration_test.

In this PR, we have not added a dedicated integration_test, which then makes the debugging and testing a lot more difficult, as you noted.

Do you know how to create a dedicated test for this? If so, go ahead and do it. If it is difficult to do, then let's not worry about it now.

**Comment by Thirumalai-Shaktivel** [comment]
@Shaikh-Ubaid is the issue https://github.com/lfortran/lfortran/issues/1953#issuecomment-1694403926 fixed?

**Comment by Shaikh-Ubaid** [comment]
>@Shaikh-Ubaid is the issue https://github.com/lfortran/lfortran/issues/1953#issuecomment-1694403926 fixed?

Yup.

**Pull request merged**